server/src/main/java/org/elasticsearch/search/aggregations/pipeline/bucketselector/BucketSelectorPipelineAggregator.java

```diff
@@ -82,34 +83,41 @@ public InternalAggregation reduce(InternalAggregation aggregation, ReduceContext
-            Map<String, Object> vars = new HashMap<>();
-            if (script.getParams() != null) {
-                vars.putAll(script.getParams());
+        if ("expression".equals(script.getLang())) {
+            BucketAggregateToDoubleScript.Factory factory =
+                reduceContext.scriptService().compile(script, BucketAggregateToDoubleScript.CONTEXT);
+            BucketAggregateToDoubleScript executableScript = factory.newInstance();
+            for (InternalMultiBucketAggregation.InternalBucket bucket : buckets) {
+                if (executableScript.execute(scriptArgs(originalAgg, bucket)) == 1.0) {
+                    newBuckets.add(bucket);
+                }
             }
-            for (Map.Entry<String, String> entry : bucketsPathsMap.entrySet()) {
-                String varName = entry.getKey();
-                String bucketsPath = entry.getValue();
-                Double value = resolveBucketValue(originalAgg, bucket, bucketsPath, gapPolicy);
-                vars.put(varName, value);
-            }
-            // TODO: can we use one instance of the script for all buckets? it should be stateless?
-            ExecutableScript executableScript = factory.newInstance(vars);
-            Object scriptReturnValue = executableScript.run();
-            final boolean keepBucket;
-            // TODO: WTF!!!!!
-            if ("expression".equals(script.getLang())) {
-                double scriptDoubleValue = (double) scriptReturnValue;
-                keepBucket = scriptDoubleValue == 1.0;
-            } else {
-                keepBucket = (boolean) scriptReturnValue;
-            }
-            if (keepBucket) {
-                newBuckets.add(bucket);
+        } else {
+            BucketAggregateToBooleanScript.Factory factory =
+                reduceContext.scriptService().compile(script, BucketAggregateToBooleanScript.CONTEXT);
+            BucketAggregateToBooleanScript executableScript = factory.newInstance();
+            for (InternalMultiBucketAggregation.InternalBucket bucket : buckets) {
+                if (executableScript.execute(scriptArgs(originalAgg, bucket))) {
+                    newBuckets.add(bucket);
+                }
             }
         }
         return originalAgg.create(newBuckets);
     }
+
+    private Map<String, Object> scriptArgs(InternalMultiBucketAggregation<InternalMultiBucketAggregation,
+        InternalMultiBucketAggregation.InternalBucket> originalAgg, InternalMultiBucketAggregation.InternalBucket bucket) {
+        Map<String, Object> vars = new HashMap<>();
+        if (script.getParams() != null) {
+            vars.putAll(script.getParams());
+        }
+        for (Map.Entry<String, String> entry : bucketsPathsMap.entrySet()) {
+            String varName = entry.getKey();
+            String bucketsPath = entry.getValue();
+            Double value = resolveBucketValue(originalAgg, bucket, bucketsPath, gapPolicy);
+            vars.put(varName, value);
```

Comment:
Putting the bucket accessors inside `params` is how we currently do this, but I don't think it is how we should do this long term. Params should be solely for user provided params (they should be read only, since they are parsed once and should be the same across all documents). Can you please open an issue to discuss with the search/aggs team how these buckets should be accessible in the future? With contexts we could potentially not use `buckets_path` at all, and have a `buckets` variable available in the script. eg:

Current:
```
"bucket_selector": {
    "buckets_path": {
        "foobar": "foo>bar"
      },
      "script": "params.foobar > 200"
}
```

Potential new script use:
```
"bucket_selector": {
    "script": "buckets['foo']['bar'].value > 200"
}
```

Reply:
Done https://github.com/elastic/elasticsearch/issues/32189 :)